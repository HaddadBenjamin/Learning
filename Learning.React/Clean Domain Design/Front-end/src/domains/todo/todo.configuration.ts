export const routes =
{
    api : '/todos'
}

export const todoReducerKey = 'todos'
export const todoSagaKey = 'todos'