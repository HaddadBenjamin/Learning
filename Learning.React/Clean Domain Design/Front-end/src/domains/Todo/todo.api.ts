import axios, { AxiosResponse } from 'axios'
import { ITodo } from './todo.model'
import {routes} from "./todo.configuration";
import {apiConfiguration, httpConfiguration} from "../../shared/shared.configuration";

export const baseUrl = `${apiConfiguration.baseUrl}${routes.api}`

export const getAllTodos = async () : Promise<AxiosResponse<ITodo[]>> =>
    (await axios.get(baseUrl, httpConfiguration.default)).data

export const addTodo = async (todo : ITodo) : Promise<AxiosResponse<ITodo>> =>
    (await axios.post(baseUrl, todo, httpConfiguration.default)).data

export const patchTitle = async (id : string, title : string) : Promise<AxiosResponse<ITodo>> => 
    (await axios.patch(`${baseUrl}/${id}`, { title : title }, httpConfiguration.default)).data

export const patchCompleted = async (id : string, completed : boolean) : Promise<AxiosResponse<ITodo>> => 
    (await axios.patch(`${baseUrl}/${id}`, { completed : completed }, httpConfiguration.default)).data

export const removeTodo = async (id : string) : Promise<AxiosResponse<ITodo>> =>
    (await axios.delete(`${baseUrl}/${id}`, httpConfiguration.default)).data