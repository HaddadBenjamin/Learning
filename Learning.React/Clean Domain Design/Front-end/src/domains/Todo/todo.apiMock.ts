import {Router} from "express";
import {todoStateMock} from "./todo.mock";
import {routes} from "./todo.configuration";
import {ITodo} from "./todo.model";

export default (router : Router) =>
{
    router.get(routes.api, (req, res) =>
    {
        res.status(200)
        res.json(todoStateMock.todos)
    })

    router.post(routes.api, (req, res) =>
    {
        const todo = req.body as ITodo

        todoStateMock.todos.push(todo)

        res.status(201)
        res.json(todo)
    })

    router.patch(`${routes.api}/:id`, (req, res) =>
    {
        const { id } = req.params
        let todo = todoStateMock.todos.find(todo => todo.id === id)

        if (todo)
            todo = { ...todo, ...req.body }

        todoStateMock.todos = todoStateMock.todos.filter(t => t.id === id ? todo : t)

        res.json(todo)
    })

    router.delete(`${routes.api}/:id`, (req, res) =>
    {
        const { id } = req.params
        const todo = todoStateMock.todos.find(todo => todo.id === id)

        if (todo)
            todoStateMock.todos = todoStateMock.todos.filter(todo => todo.id !== id)

        res.status(200)
        res.json(todo)
    })
}