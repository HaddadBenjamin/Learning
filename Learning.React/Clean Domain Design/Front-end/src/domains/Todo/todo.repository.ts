import {newGuid} from "shared/utils/newGuid";
import {ITodo} from "./todo.model";

export const createTodo = (title : string) : ITodo =>
({
    id : newGuid(),
    completed : false,
    title
})