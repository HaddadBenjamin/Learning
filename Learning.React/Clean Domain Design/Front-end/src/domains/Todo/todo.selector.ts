import { ApplicationState } from '../Root/root.state';
import { TodoState } from './todo.state';

export const selectTodos = (state : ApplicationState) : TodoState  => state.todos