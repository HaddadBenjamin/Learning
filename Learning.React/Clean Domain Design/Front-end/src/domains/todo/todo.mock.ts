import {TodoState} from "./todo.state";
import {ActionStatus} from "shared/domains/redux/redux.model";

export const emptyTodoStateMock : TodoState =
{
    status : ActionStatus.Loading,
    error : undefined,
    todos : []
}

export const todoStateMock : TodoState =
{
    ...emptyTodoStateMock,
    todos : [
        {
            id : "1",
            title : "Faire le m√©nage",
            completed : false,
        },
        {
            id : "2",
            title : "Faire ma compta",
            completed : true,
        },
        {
            id : "3",
            title : "Appeler Marine",
            completed : false,
        },
        {
            id : "4",
            title : "Acheter des croquettes pour petit chien",
            completed : false,
        },
    ]
}

export const todoMock = todoStateMock.todos[0]
export const {title : todoTitleMock, id : todoIdMock} = todoStateMock.todos[1]
export const errorMessageMock = '404 not found'
export const errorMock  = new Error(errorMessageMock)