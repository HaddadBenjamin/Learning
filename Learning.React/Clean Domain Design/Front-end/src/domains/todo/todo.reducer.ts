import { TodoState, initialTodoState } from './todo.state';
import { TodoAction, TodoActions } from './todo.action';
import { failedActionMetadata, loadedActionMetadata, loadingActionMetadata } from 'shared/domains/redux/redux.util';

export const todoReducer = (state: TodoState = initialTodoState, action: TodoActions)
: TodoState =>
{
    switch (action.type)
    {
        case TodoAction.GET_TODOS_REQUEST :     return loadingActionMetadata(state)
        case TodoAction.GET_TODOS_SUCCESS :     return { ...loadedActionMetadata(state), todos: action.payload }
        case TodoAction.GET_TODOS_FAILED :      return failedActionMetadata(action.error, state)

        case TodoAction.ADD_TODO_REQUEST :      return loadingActionMetadata(state)
        case TodoAction.ADD_TODO_SUCCESS :      return loadedActionMetadata(state)
        case TodoAction.ADD_TODO_FAILED :       return failedActionMetadata(action.error, state)
     
        case TodoAction.EDIT_TODO_REQUEST :     return loadingActionMetadata(state)
        case TodoAction.EDIT_TODO_SUCCESS :     return loadedActionMetadata(state)
        case TodoAction.EDIT_TODO_FAILED :      return failedActionMetadata(action.error, state)

        case TodoAction.TOGGLE_TODO_REQUEST :   return loadingActionMetadata(state)
        case TodoAction.TOGGLE_TODO_SUCCESS :   return loadedActionMetadata(state)
        case TodoAction.TOGGLE_TODO_FAILED :    return failedActionMetadata(action.error, state)

        case TodoAction.REMOVE_TODO_REQUEST :   return loadingActionMetadata(state)
        case TodoAction.REMOVE_TODO_SUCCESS :   return loadedActionMetadata(state)
        case TodoAction.REMOVE_TODO_FAILED :    return failedActionMetadata(action.error, state)

        default : return state
    }
}