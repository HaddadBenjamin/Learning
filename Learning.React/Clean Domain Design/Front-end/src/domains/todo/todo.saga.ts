import { call, put, select, takeLatest } from 'redux-saga/effects'
import {
   addTodoFailedAction,
   AddTodoRequestAction,
   addTodoSuccessAction,
   editTodoFailedAction,
   EditTodoRequestAction,
   editTodoSuccessAction,
   getTodosFailedAction, getTodosRequestAction,
   getTodosSuccessAction,
   removeTodoFailedAction,
   RemoveTodoRequestAction,
   removeTodoSuccessAction,
   TodoAction,
   toggleTodoFailedAction,
   ToggleTodoRequestAction,
   toggleTodoSuccessAction
} from './todo.action';
import {
   addTodo,
   getAllTodos,
   patchTitle,
   removeTodo,
   patchCompleted
} from './todo.api';
import { ITodo } from './todo.model';
import { selectTodos } from './todo.selector';
import { TodoState } from './todo.state';
import {createTodo} from "./todo.repository";

export function* getAllTodosSaga()
{
   try
   {
      const todos : ITodo[] = yield call(getAllTodos);
      yield put(getTodosSuccessAction(todos));
   }
   catch (error)
   {
      yield put(getTodosFailedAction(error.message));
   }
}

export function* addTodoSaga(action : AddTodoRequestAction)
{
   try {
      const newTodo: ITodo = yield call(addTodo, createTodo(action.payload.title));

      yield put(addTodoSuccessAction(newTodo));
      yield put(getTodosRequestAction());
   }
   catch (error)
   {
      yield put(addTodoFailedAction(error.message));
   }
}

export function* editTodoSaga(action : EditTodoRequestAction)
{
   try
   {
      const { id, title } = action.payload
      const editedTodo : ITodo = yield call(patchTitle, id, title);

      yield put(editTodoSuccessAction(editedTodo));
      yield put(getTodosRequestAction());
   }
   catch (error)
   {
      yield put(editTodoFailedAction(error.message));
   }
}

export function* toggleTodoSaga(action : ToggleTodoRequestAction)
{
   try
   {
      const { id } = action.payload      
      const todoState : TodoState = yield select(selectTodos);
      const todo : ITodo = todoState.todos.find(t => t.id === id) as ITodo
      const toggledTodo : ITodo = { ...todo, completed : !todo.completed}
   
      yield call(patchCompleted, id, toggledTodo.completed);
      yield put(toggleTodoSuccessAction(toggledTodo));
      yield put(getTodosRequestAction());
   }
   catch (error)
   {
      yield put(toggleTodoFailedAction(error.message));
   }
}

export function* removeTodoSaga(action : RemoveTodoRequestAction)
{
   try
   {
     const { id } = action.payload

     yield call(removeTodo, id);
     yield put(removeTodoSuccessAction(id));
      yield put(getTodosRequestAction());
   }
   catch (error)
   {
      yield put(removeTodoFailedAction(error.message));
   }
}

function* watchTodosSagas() {
  yield takeLatest(TodoAction.GET_TODOS_REQUEST, getAllTodosSaga);
  yield takeLatest(TodoAction.ADD_TODO_REQUEST, addTodoSaga);
  yield takeLatest(TodoAction.EDIT_TODO_REQUEST, editTodoSaga);
  yield takeLatest(TodoAction.TOGGLE_TODO_REQUEST, toggleTodoSaga);
  yield takeLatest(TodoAction.REMOVE_TODO_REQUEST, removeTodoSaga);
}

export default watchTodosSagas;