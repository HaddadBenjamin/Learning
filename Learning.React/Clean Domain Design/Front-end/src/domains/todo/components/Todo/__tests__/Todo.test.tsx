import {fireEvent, render} from "@testing-library/react";
import Todo from "../Todo";
import {useDispatch} from "react-redux";
import {todoMock, todoTitleMock} from "domains/todo/todo.mock";
import {editTodoRequestAction, removeTodoRequestAction, toggleTodoRequestAction} from "domains/todo/todo.action";

jest.mock('react-redux')
const mockUseDispatch = useDispatch as jest.MockedFunction<typeof useDispatch>
const mockDispatch = jest.fn()

describe("Todo", () => {
    beforeEach(() => {
        (mockUseDispatch as jest.Mock).mockReturnValue(mockDispatch)
    })

    it("should match snapshot", () => {
        // Given & When
        const {container} = render(<Todo {...todoMock} />)

        // Then
        expect(container).toMatchSnapshot()
    })

    it("should dispatch editTodoRequestAction when title change", () =>
    {
        // Given
        const {container} = render(<Todo {...todoMock} />)
        const input = container.getElementsByTagName('input')[1]

        // When
        fireEvent.change(input, {target: {value: todoTitleMock}})

        // Then
        expect(mockDispatch).toBeCalledWith(editTodoRequestAction(todoMock.id, todoTitleMock))
    })

    it("should dispatch toggleTodoRequestAction when toggle is check", () =>
    {
        // Given
        const {container} = render(<Todo {...todoMock} />)
        const toggle = container.getElementsByTagName('input')[0]

        // When
        fireEvent.click(toggle)

        // Then
        expect(mockDispatch).toBeCalledWith(toggleTodoRequestAction(todoMock.id))
    })

    it("should dispatch removeTodoRequestAction when remove button is clicked", () =>
    {
        // Given
        const {getByText} = render(<Todo {...todoMock} />)
        const removeButton = getByText('X')

        // When
        fireEvent.click(removeButton)

        // Then
        expect(mockDispatch).toBeCalledWith(removeTodoRequestAction(todoMock.id))
    })
})