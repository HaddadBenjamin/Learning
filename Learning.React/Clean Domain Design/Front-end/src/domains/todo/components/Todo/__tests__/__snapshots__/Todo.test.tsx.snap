// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo should match snapshot 1`] = `
<div>
  <div
    class="todo"
  >
    <div>
      <input
        class="toggle"
        type="checkbox"
      />
      <input
        class="content"
        value="Faire le mÃ©nage"
      />
    </div>
    <div
      class="removeButton"
    >
      X
    </div>
  </div>
</div>
`;
